### ISSUE TITLE
[feature suggestion] Gemfile overrides gem dependencies, pessimistic dependency resolver

### ISSUE THREAD

**Issue by ajgrover** [bundler/bundler-features#82]
Hi all - 

@sxl1092, @bgott, @rrperez and I are 4 Stanford students working on Rails through [Facebook OpenAcademy](https://www.facebook.com/notes/facebook-engineering/facebook-open-academy-bringing-open-source-to-cs-curricula/10151806121378920) for our respective senior projects for school. As our project, we were given the suggestion to propose 2 new features to Bundler that would be useful to the Rails team. We are exploring if the best way to do this would be through code merged into the Bundler core or as a plugin but wanted to get your feedback before we started. 

First off, it would be useful to have a pessimistic dependency resolver: something like `bundle install --use-oldest` that would install the oldest possible version of all gems. This would allow us to test to ensure that we are not accidentally using newer versions of gems that may not be compatible with those declared in the Gemfile. That way, we would know and be able to bump the version declared in the Gemfile to match ones that will not break existing code.

Secondly, the ability for a Gemfile to override conflicting dependencies of a gem it is installing. For example, the activesupport gem depends on i18n '~> 0.7'. That means Bundler will refuse to continue if we have another gem, which depends on (a theoretical) i18n 1.0 (because '~> 0.7' is equivalent to '>= 0.7, < 1.0').

While that's reasonable in general (the newer i18n may indeed be incompatible), it's a pain for testing. (Or, if the new version actually is compatible for our purposes, if the maintainer of activesupport is slow in publishing a new release that declares 1.0 okay.)

So, it would be nice if we could add something to our application's Gemfile, which would tell Bundler that we, as the application author, acknowledge that we're asking it to install two incompatible gems, and we'll take responsibility for the consequences.

```ruby
override_dependency 'activesupport', 'i18n', '>= 0.7, < 2.0'
```

Thanks in advance for listening to our suggestions and we look forward to hearing your advice on the best way to proceed.

**Comment by indirect** [comment]
Hi Alex. I'm not entirely clear on your explanation of why you want the "use oldest" feature. Can you give a concrete example of a situation where you would use it?

Forcibly overriding dependencies is a feature that gets asked for on occasion, but so far we have decided against it every time it has come up. It adds complexity, indirection, and "magic" to the way that gemspecs work. It's also (relatively) straightforward to fork a gem repo, edit the gemspec, and then use that git repo while waiting for a gem author to release an updated gem, and so we have recommended that instead.

**Comment by ajgrover** [comment]
Got it. That makes sense.

As for the use-oldest feature, bundler's resolver finds the newest set of gems that satisfy a given Gemfile's requirements. This means we can sometimes end up with a lie in the Gemfile: if we say that we require gem foo '>= 1.0' when we start work on our project, that's fine.

But as time goes on, the developers will actually have foo 1.30 (or whatever) installed. So nobody notices when we start calling a method that was only added in foo 1.12: it's now possible to have a version of the gem installed that meets our declared dependency, but isn't actually compatible with our code.

If we could run our test suite after a bundle install --use-oldest, then we could be sure that our code really does work with the oldest version it claims. CI would fail as soon as we started calling the too-new method, so we would know we'd broken compatibility with foo 1.0, and either use a more backwards-compatible implementation of the functionality, or determine which version introduced the feature we need, and bump our dependency to match.

**Comment by indirect** [comment]
That makes sense. It's probably harder to implement than it initially sounds, but that sounds like a valuable feature to add to core to me. Is your team interested in implementing it?

**Comment by segiddins** [comment]
I think implementing `use-oldest` would just require reversing the specs that are passed into molinillo to be oldest-first, but beware, you're still not testing anywhere _near_ the entire permutation space for your requirements by doing so.

**Comment by arthurnn** [comment]
cc @matthewd

**Comment by ajgrover** [comment]
Hey all - sorry for the delay. We have been looking into this and should be ready to make a PR in the next few days. I'm not sure it would be necessary to test the entire permutation space if we are basing the use case for this feature simply on new functionality in dependencies. It would be possible to test every permutation but we are moving forward based on the assumption that we just want the oldest possible as that is simpler and achieves the behavior desired.

**Comment by indirect** [comment]
Sounds good. Thanks for the update!

**Comment by wjordan** [comment]
See also #20 for "Gemfile overrides gem dependencies" feature discussion.
